Interrupt stack size unification + test

mprse opened the pull request:
### Description

This PR adds test which verifies sizes of: ISR stack, main thread stack, default user thread stack and updates linker scripts for `ARM`, `GCC_ARM` and `IAR` compilers to unify interrupt/boot stack size using framework provided in PR https://github.com/ARMmbed/mbed-os/pull/8039.

Tested on the following boards:
```
K66F
K22F
EV_COG_AD4050LZ
NRF52_DK
NRF51_DK
NRF52840_DK
DISCO_L475VG_IOT01A
DISCO_F413ZH
NUCLEO_L476RG
NUCLEO_F207ZG
NUCLEO_F429ZI
LPC4088
FF_LPC546XX
EFM32GG_STK3700
```

### Pull request type

<!-- 
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [X] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

mprse committed:
Add stack size unification test

Test checks stack sizes:
 - ISR stack: exactly 1KB (with exception for NORDIC - 2KB and RENESAS - Cortex A targets not supported for this test)
 - Main thread stack: exactly 4KB (with some exceptions - 3KB)
 - Thread stack: exactly 4KB

ciarmcom commented:
@mprse, thank you for your changes.
@ARMmbed/mbed-os-core @ARMmbed/mbed-os-hal @ARMmbed/mbed-os-test @ARMmbed/mbed-os-maintainers please review.

jeromecoutant commented:
Hi
First quick test with L073:

| target                | platform_name | test suite                               | result | elapsed_time (sec) | copy_method |
|-----------------------|---------------|------------------------------------------|--------|--------------------|-------------|
| NUCLEO_L073RZ-ARM     | NUCLEO_L073RZ | tests-mbed_hal-stack_size_unification    | OK     | 18.91              | default     |
| NUCLEO_L073RZ-ARM     | NUCLEO_L073RZ | tests-mbedmicro-rtos-mbed-heap_and_stack | FAIL   | 19.96              | default     |
| NUCLEO_L073RZ-GCC_ARM | NUCLEO_L073RZ | tests-mbed_hal-stack_size_unification    | OK     | 21.21              | default     |
| NUCLEO_L073RZ-GCC_ARM | NUCLEO_L073RZ | tests-mbedmicro-rtos-mbed-heap_and_stack | OK     | 22.09              | default     |
| NUCLEO_L073RZ-IAR     | NUCLEO_L073RZ | tests-mbed_hal-stack_size_unification    | OK     | 18.84              | default     |
| NUCLEO_L073RZ-IAR     | NUCLEO_L073RZ | tests-mbedmicro-rtos-mbed-heap_and_stack | OK     | 19.62              | default     |

Please check the failed test (OK in master branch)

mprse commented:
@jeromecoutant Thanks for checking this change. Unfortunately I don't have `NUCLEO_L073RZ`. Could you provide the test output?

deepikabhavnani commented:
@mprse  - Will appreciate if you can split the PR into multiple based on targets (not necessary one per target but group of them). So many files have changed that even drop down button for  files selection is not working, its extremely difficult to review.

jeromecoutant commented:
@mprse it seems that this test is FAIL for all the targets only with ARM

mprse commented:
> @mprse it seems that this test is FAIL for all the targets only with ARM

I pushed some fixes, but I was not able to test it well today. Tested only on NUCLEO_L476RG.

mprse commented:
> @mprse it seems that this test is FAIL for all the targets only with ARM

Few issues have been encountered while analysing the failing test. The first problem was that for the following targets isr/boot stack has been located in `RW_IRAM2` instead `RW_IRAM1`: 
```
targets/TARGET_STM/TARGET_STM32F4/TARGET_STM32F437xG/device/TOOLCHAIN_ARM_STD/stm32f437xx.sct
targets/TARGET_STM/TARGET_STM32F4/TARGET_STM32F439xI/device/TOOLCHAIN_ARM_STD/stm32f439xx.sct
targets/TARGET_STM/TARGET_STM32L4/TARGET_MTS_DRAGONFLY_L471QG/device/TOOLCHAIN_ARM_STD/stm32l471xx.sct
targets/TARGET_STM/TARGET_STM32L4/TARGET_STM32L443xC/device/TOOLCHAIN_ARM_STD/stm32l443xx.sct
targets/TARGET_STM/TARGET_STM32L4/TARGET_STM32L475xG/device/TOOLCHAIN_ARM_STD/stm32l475xx.sct
targets/TARGET_STM/TARGET_STM32L4/TARGET_STM32L486xG/device/TOOLCHAIN_ARM_STD/stm32l486xx.sct
targets/TARGET_STM/TARGET_STM32L4/TARGET_STM32L476xG/device/TOOLCHAIN_ARM_STD/stm32l476xx.sct
```
The problem has been fixed, but then I checked heap and stack locations and found another problem:
```
mbed_stack_isr_start: 0x20017C00
mbed_stack_isr_end:   0x20018000     <---- mbed_stack_isr_end == mbed_heap_end !!!
mbed_stack_isr_size:  0x400
mbed_heap_start:      0x200027F8
mbed_heap_end:        0x20018000     <---- mbed_stack_isr_end == mbed_heap_end !!!
mbed_heap_size:       0x15808

```
Isr/boot stack was located inside the heap (`mbed_stack_isr_end == mbed_heap_end`). Which is not good. It looks like the problem is in `mbed_boot_arm_std.c` file:

https://github.com/ARMmbed/mbed-os/blob/463a4536e24135d19346093da5e39d84d6c96db9/rtos/TARGET_CORTEX/TOOLCHAIN_ARM_STD/mbed_boot_arm_std.c#L30-L75

This PR adds definition of `ARM_LIB_STACK` to all scatter files and based on `ARM_LIB_STACK` defines `ISR_STACK_START` and `ISR_STACK_SIZE` (https://github.com/ARMmbed/mbed-os/pull/9092/commits/40ff7f1adf1874806814898ee6209d3b85cd3e32). 
This means that `__rt_entry()` function will use now explicitly specified interrupt stack (line 58 in above code snippet). It looks like the problem is now with heap definition. Heap is defined based on `INITIAL_SP` which points to the top of the RAM (line 34). Heap defined in this way actually defines heap+stack region - which is ok for case when `ISR_STACK_START` is not specified, because interrupt/boot stack will be reserved at the end of the free block. In case when `ISR_STACK_START` is specified heap definition should be changed to define only real heap size. Commit https://github.com/ARMmbed/mbed-os/pull/9092/commits/f40bb4542228def6489a5bd2963131b7b3c63664 fixes the problem.

Final results:
```
mbed_stack_isr_start: 0x20017C00     <--- stack_isr_start == mbed_heap_end 
mbed_stack_isr_end:   0x20018000
mbed_stack_isr_size:  0x400
mbed_heap_start:      0x200027F8
mbed_heap_end:        0x20017C00     <--- stack_isr_start == mbed_heap_end 
mbed_heap_size:       0x15408
```

Now isr/boot stack is located after heap and isr/boot stack size is 1KB - as expected.

@c1728p9 Could you check if above modifications make sense?

mprse commented:
> @mprse - Will appreciate if you can split the PR into multiple based on targets (not necessary one per target but group of them). So many files have changed that even drop down button for files selection is not working, its extremely difficult to review.

Can I leave one PR and create commit for each vendor?

0xc0170 commented:
> Can I leave one PR and create commit for each vendor?

That should help and also simplify rebases as targets are changing constantly

mprse committed:
[ARM_FM] Support boot stack size configuration option

mprse committed:
[ARM_SSG] Support boot stack size configuration option

mprse committed:
[Analog_Devices] Support boot stack size configuration option

mprse committed:
[Atmel] Support boot stack size configuration option

mprse committed:
[Cypress] Support boot stack size configuration option

mprse committed:
[Freescale] Support boot stack size configuration option

mprse committed:
[Maxim] Support boot stack size configuration option

mprse committed:
[NXP] Support boot stack size configuration option

mprse committed:
[ONSEMI] Support boot stack size configuration option

mprse committed:
[RDA] Support boot stack size configuration option

mprse committed:
[Realtek] Support boot stack size configuration option

mprse committed:
[STM] Support boot stack size configuration option

mprse committed:
[Silicon_Labs] Support boot stack size configuration option

mprse committed:
[TOSHIBA] Support boot stack size configuration option

mprse committed:
[WIZNET] Support boot stack size configuration option

mprse committed:
[ublox] Support boot stack size configuration option

deepikabhavnani commented:
> Can I leave one PR and create commit for each vendor?

Commit per vendor will help with the review as well. 
From my past experience with changing linker files, it was worth splitting the PR's between vendors. Avoided rebase few times and some got merged pretty fast. Your call to have separate commits or PR.

c1728p9 commented:
Hi @mprse, your modifications make sense. I think you'll need to make the same modifications for GCC.

With this change will all targets be setting heap/stack info through their linker script? Once that is the case we should remove the configurable macro values HEAP_START,HEAP_SIZE,ISR_STACK_START,ISR_STACK_SIZE and use the linker symbols directly in the respective toolchain boot file - `"mbed_boot_<toolchain>.c"`. That may be outside the scope of this PR though.

0xc0170 commented:
CI started to get early errors. Still needs reviews

mprse commented:
> With this change will all targets be setting heap/stack info through their linker script? Once that is the case we should remove the configurable macro values HEAP_START,HEAP_SIZE,ISR_STACK_START,ISR_STACK_SIZE and use the linker symbols directly in the respective toolchain boot file - "mbed_boot_<toolchain>.c". That may be outside the scope of this PR though.

@c1728p9 Thanks for checking this changes. Yes, all targets should set heap/stack info through their linker scripts. I tried to address your proposal and use the linker symbols directly in the respective toolchain boot file (https://github.com/ARMmbed/mbed-os/pull/9092/commits/dc43afc836b5a2af99c60f63a793f4b2f846f19a) and remove HEAP_START,HEAP_SIZE,ISR_STACK_START,ISR_STACK_SIZE macros(https://github.com/ARMmbed/mbed-os/pull/9092/commits/c6cfa96ab9ac5605b17800a01c67a74690a3d91b). 

At the moment I have the following problems:
- I need to check how to update Renesas linker scripts which are for Cortex-A cores.
- @jeromecoutant cases where there are two memory banks like DISCO_L475VG_IOT01A. Please see the follwing example:

https://github.com/ARMmbed/mbed-os/blob/995725b50a7538e3f0b331823c0bd40a5691d9a5/targets/TARGET_STM/TARGET_STM32L4/TARGET_STM32L475xG/device/TOOLCHAIN_GCC_ARM/STM32L475XX.ld#L151-L170

Heap is located in SRAM1 and stack in SRAM2. This causes 'tests-mbedmicro-rtos-mbed-heap_and_stack' to fail (Heap in wrong location). This is because heap start and size is now defined based on `__end__` and `__StackLimit` symbols which are in two different memory banks. The solution would be to place heap and stack in one memory bank. Is it ok?

BTW why this is done in such way? I thought that memory layout assumes that stack is placed after the heap.

deepikabhavnani commented:
> The solution would be to place heap and stack in one memory bank. Is it ok?

I don't think this will be ok. We do not support heap split between multiple banks and as a result one bank is either not utilized or partially used. 
In case of multiple banks it is advised to have heap in separate bank. Can we update the test accordingly?

Multiple issues related to HEAP split + HEAP waste in GCC (allocation in 4K chunks) are reported and solution was to split HEAP / Stack in separate sections, latest was this:

https://github.com/ARMmbed/mbed-os/issues/8963

> I thought that memory layout assumes that stack is placed after the heap.

We actually need to rethink on this layout, it should be default suggested but not forced. This layout was adopted for mbed 2 (4K stack) and is good for small devices. 
But in case of Mbed , ISR stack is 1K and placing it at the end (after heap) results in 3K RAM not used for GCC (Heap allocation should be 4K aligned)

mprse commented:
@deepikabhavnani Many thanks for the clarification. It was very helpful for me. 

> Can we update the test accordingly?

There is no problem with the test since the test uses `mbed_heap_start`, `mbed_heap_size `, `mbed_stack_isr_start` and `mbed_stack_isr_size` symbols. In current version definition of `mbed_heap_size ` is invalid for cases when isr stack is not placed after heap:
```
mbed_heap_start = ((unsigned char*)__end__);
mbed_heap_size = ((uint32_t)&__StackLimit) - (uint32_t)mbed_heap_start;
```
The solution would be to define `__HeapLimit ` in all ld files and define heap size as follows:
```
mbed_heap_size = ((uint32_t)&__HeapLimit) - (uint32_t)mbed_heap_start;
```

Since the PR is heavy enough I will make a step back and restore HEAP_START, HEAP_SIZE, ISR_STACK_START, ISR_STACK_SIZE macros definitions. In the next step we need to provide `__HeapLimit` definition in all GCC linker scripts and then we should be able to remove mentioned macros and use the linker symbols directly in the respective toolchain boot files.

mprse committed:
Define heap/stack start and size based on linker script symbols for ARM and GCC_ARM

mprse commented:
I have reverted the last changes(restored HEAP_START, HEAP_SIZE, ISR_STACK_START, ISR_STACK_SIZE macros definitions).

@jeromecoutant I have successfully run both tests on the following STM boards:
`NUCLEO_F070RB, NUCLEO_F207ZG,NUCLEO_F429ZI,NUCLEO_L476RG,DISCO_L475VG_IOT01A`.

@deepikabhavnani I have found one issue for `NUCLEO_F070RB`. Main thread stack size is equal 3KB and we expect 4KB. I saw that this was recently changed, this board has only 16 KB of RAM, so I added exception for this board in the test.

@0xc0170 Can we try to run the test run again?

0xc0170 commented:
Please fix astyle, we will trigger CI

mprse commented:
> Please fix astyle, we will trigger CI

astyle has passed!

0xc0170 commented:
CI started

deepikabhavnani commented:
>The solution would be to define __HeapLimit in all ld files and define heap size as follows:

Sounds good to me 👍 

>  I have found one issue for NUCLEO_F070RB. Main thread stack size is equal 3KB and we expect 4KB. I saw that this was recently changed, this board has only 16 KB of RAM, so I added exception for this board in the test.

Few more targets to this list: TARGET_NANO100, TARGET_STM32F070RB, TARGET_STM32F072RB, TARGET_TMPM066 and TARGET_TMPM46B. Reference: https://github.com/ARMmbed/mbed-os/pull/5285

mprse commented:
> Test run: FAILED

Looks like I missed Nordic while splitting commits and problem with REALTEK_RTL8195AM_GCC_ARM has been detected.

mprse committed:
[NORDIC] Support boot stack size configuration option

mprse committed:
[GigaDevice] Support boot stack size configuration option

mprse commented:
Added the following changes:
- Rebased,
- Extended list of targets with main thread stack set to 3KB in the test,
- Updated Nordic linker scripts and tested on NRF_51_DK, NRF_52DK and NRF52840_DK boards.
- Fixed REALTEK_RTL8195AM build failure on GCC,
- Updated GigaTarget linker scripts (new) and tested if Mbed can be successfully build.

@0xc0170 Ready for the next CI round.

0xc0170 commented:
CI started

mprse commented:
> The command "W: Failed to fetch https://packagecloud.io/github/git-lfs/ubuntu/dists/trusty/InRelease Operation too slow. Less than 10 bytes/sec transferred the last 10 seconds" failed. Retrying, 2 of 3.

It looks like there was some communication problem.

mprse commented:
> Test run: FAILED

One problem left: `TMPM4G9/ARM` build failure. Typo in linker script caused the problem.

@0xc0170 Fix has been pushed. 🤞

cmonr commented:
CI started

NirSonnenschein commented:
Hi @mprse
here is a snippet from the failure logs (https://mbed-os.mbedcloudtesting.com/job/mbed-os-ci_greentea-test/646/testReport/junit/EV_COG_AD3029LZ-ARM/tests-mbedmicro-rtos-mbed-heap_and_stack/EV_COG_AD3029LZ_armcc___Test_isr_stack_in_range/ for and arbitrary): 
[1545428149.47][CONN][RXD] >>> Running case #3: 'Test isr stack in range'...
[1545428149.47][CONN][INF] found KV pair in stream: {{__testcase_start;Test isr stack in range}}, queued...
[1545428149.57][CONN][RXD] :209::FAIL: Interrupt stack in wrong location
[1545428149.67][CONN][INF] found KV pair in stream: {{__testcase_finish;Test isr stack in range;0;1}}, queued...
[1545428149.77][CONN][RXD] >>> 'Test isr stack in range': 0 passed, 1 failed with reason 'Assertion Failed'

this looks like a new test is failing, could this be the case?

mprse commented:
Hi @NirSonnenschein! Yes, there must be some problem with scattter files for Analog Devices boards. I need to check that.

mprse commented:
Pushed scatter file fix for `EV_COG_AD3029LZ`. @0xc0170 @cmonr @NirSonnenschein Can we try again with CI?

NirSonnenschein commented:
Hi @mprse , there is currently an issue with the arm compiler license blocking CI, we can re-run when this is resolved (hopefully very soon).

mprse commented:
>Hi @mprse , there is currently an issue with the arm compiler license blocking CI, we can re-run when this is resolved (hopefully very soon).

@NirSonnenschein @0xc0170 Please run CI if the problem has been solved.

0xc0170 commented:
CI started

mprse committed:
[TT] Support boot stack size configuration option

mprse commented:
> Test run: FAILED

New target has been added: `TARGET_TT`.
Rebased and adapted `TT_M3HQ` linker script files. 
Successfully build using all compilers.

@0xc0170 Ready for CI.

0xc0170 commented:
CI restarted

cmonr commented:
@mbed-os-maintainers @ARMmbed/mbed-os-core @ARMmbed/mbed-os-hal @ARMmbed/mbed-os-test-team When y'all have the chance, this appears ready to go.

deepikabhavnani reviewed (approved):
Looks good to me 👍

cmonr commented:
I noticed a couple of `#if !defined(MBED_BOOT_STACK_SIZE)` in some linker files. What would be the reason that this would not already exist?

I also noticed that the above is sparingly in linker files, but all over the place in the scatter files. Was this intentional?

Would also appreciate a quick look by @ARMmbed/mbed-os-ipcore since there are a bunch of small netsocket changes.

The last thing. This _might_ need a rebase, if ony because I recall a PR disabling the MPU for the NRF52 recently coming in, but this diff appears to re-enable it in targets.json... Otherwise, LGTM!

mprse commented:
> I noticed a couple of #if !defined(MBED_BOOT_STACK_SIZE) in some linker files. What would be the reason that this would not already exist?
> 
> I also noticed that the above is sparingly in linker files, but all over the place in the scatter files. Was this intentional?

According to my understanding currently `MBED_BOOT_STACK_SIZE` should be defined for Mbed 2 and Mbed 5 and all devices. So it should not be a problem if `#if !defined(MBED_BOOT_STACK_SIZE)` is skipped, but this check has been added to `*.ld`, `*.icf` and `*.sct` files for consistency with the example provided with the framework for configuring boot stack size( https://github.com/ARMmbed/mbed-os/pull/8039/commits/66cce67d38fdae3d4f395265f10c302d63824cb4).

SeppoTakalo reviewed (approved):
I did not spot any network test related changes, therefore I can approve from IP Core team point of view.

0xc0170 commented on TESTS/mbed_hal/stack_size_unification/stack_size_unification.h:
```diff
@@ -0,0 +1,49 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2017-2018 ARM Limited
```
if we add new files, we should add SPDX identifier as well, can you please

mprse replied:
Fixed!

0xc0170 reviewed (changes requested):
LGTM, just 2 small edits

mprse committed:
Add SPDX-License-Identifier and Copyright in new files

jamesbeyond reviewed (approved):
Looks good, well Done!

mprse commented:
Rebased to resolve conflicts.

0xc0170 commented:
CI started

mprse commented:
> Test run: FAILED

Looks like new target (`NUCLEO_H743ZI`) has been added meantime.
Adapted linker scripts and tested if Mbed can be build after the changes on `NUCLEO_H743ZI`/`all compilers`.
Ready for the next, hopefully final, CI round.

0xc0170 commented:
CI restarted

mprse commented:
> The command "W: Failed to fetch https://packagecloud.io/github/git-lfs/ubuntu/dists/trusty/InRelease The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 6B05F25D762E3157" failed. Retrying, 2 of 3.

Looks like some communication issue reported by Travis.

0xc0170 commented:
Travis restarted

mprse commented:
Looks like the same error.

0xc0170 commented:
Travis having hiccups today

mprse commented:
> Travis having hiccups today

Poor Travis 😢

0xc0170 commented:
I found a workaround , #9289

0xc0170 commented:
Just travis (they might fix it meanwhile). 

@mprse What is missing for this one to land ? Will it have impact on following patch releases (as it touches lot of targets) - it should not, as changes there are simple. This should land soon to have any new target to comply with this

mprse commented:
> What is missing for this one to land ?

It looks like at this moment its ready (review and CI passed - almost).

> Will it have impact on following patch releases?

Hopefully not... If I didn't mess something (lots of changes). CI proved that Mbed can be successfully build with these changes for all targets/compilers and the change has no impact on CI test targets. 

> This should land soon to have any new target to comply with this

Agree.

0xc0170 commented:
Travis was fixed, if we rebase, it should pass (will require new test round)

mprse commented:
> Travis was fixed, if we rebase, it should pass (will require new test round)

Done!

0xc0170 commented:
CI started

mprse commented:
> Summary: 1 of 11 test jobs failed
> Build number : 9
> Build artifacts
> 
> Failed test jobs:
> 
> jenkins-ci/mbed-os-ci_greentea-test

Seems that greentea-test has passed (no failures).

0xc0170 commented:
@ARMmbed/mbed-os-maintainers Any objections for this one landing soon?
